{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "28099295-ccd5-440f-ab0f-0fc3ae9c78d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(\"/Volumes/Extreme SSD/HP/read1.UMI.fastq\")\n",
    "r = open(\"/Volumes/Extreme SSD/HP/read2.UMI.fastq\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "779d14db-948b-43a0-9b4b-1bda75d9820e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">1_1\n",
      "\"shell_port\": 50479,\n",
      ">2_1\n",
      "\"stdin_port\": 50481,\n",
      ">3_1\n",
      "\"hb_port\": 50482,\n",
      ">4_1\n",
      "\"key\": \"d2d255b3-55031f7a06b59b37a89c7e1e\",\n",
      ">5_1\n",
      "\"signature_scheme\": \"hmac-sha256\",\n",
      ">6_1\n",
      "\"jupyter_session\": \"/Volumes/Extreme SSD/HP/Scripts/Eliminating_Duplicates_for_Midway.ipynb\"\n"
     ]
    }
   ],
   "source": [
    "__author__\t\t= \"Sander Granneman\"\n",
    "__copyright__\t= \"Copyright \"\n",
    "__version__\t\t= \"0.0.5\"\n",
    "__credits__\t\t= [\"Sander Granneman\"]\n",
    "__maintainer__\t= \"Sander Granneman\"\n",
    "__email__\t\t= \"sgrannem@ed.ac.uk\"\n",
    "__status__\t\t= \"Production\"\n",
    "\n",
    "##################################################################################\n",
    "#\n",
    "#\tpyFastqDuplicateRemover.py\n",
    "#\n",
    "#\n",
    "#\tCopyright (c) Sander Granneman 2019\n",
    "#\t\n",
    "#\tPermission is hereby granted, free of charge, to any person obtaining a copy\n",
    "#\tof this software and associated documentation files (the \"Software\"), to deal\n",
    "#\tin the Software without restriction, including without limitation the rights\n",
    "#\tto use, copy, modify, merge, publish, distribute, sublicense, and/or sell\n",
    "#\tcopies of the Software, and to permit persons to whom the Software is\n",
    "#\tfurnished to do so, subject to the following conditions:\n",
    "#\t\n",
    "#\tThe above copyright notice and this permission notice shall be included in\n",
    "#\tall copies or substantial portions of the Software.\n",
    "#\t\n",
    "#\tTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\n",
    "#\tIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\n",
    "#\tFITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\n",
    "#\tAUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\n",
    "#\tLIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,\n",
    "#\tOUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN\n",
    "#\tTHE SOFTWARE.\n",
    "#\n",
    "##################################################################################\n",
    "\n",
    "\n",
    "import sys\n",
    "import re\n",
    "import six\n",
    "from collections import defaultdict\n",
    "from optparse import *\n",
    "\n",
    "def getfilename(filepath):\n",
    "\t\"\"\" Returns the file name without path and extension\"\"\"\n",
    "\treturn filepath.split(\"/\")[-1].split(\".\")[0]\n",
    "\n",
    "def trackDuplicates(file_input):\n",
    "\t\"\"\" keeps track of the cDNA duplicates using both random barcode sequences (if available)\n",
    "\tand the nucleotide sequence. Counts the number of duplicates. Returns a dictionary \"\"\"\n",
    "\trandbarcode = None\n",
    "\theader\t\t= str()\n",
    "\ttracker\t\t= defaultdict(int)\n",
    "\tsplitter\t= \"_\"\t# the character(s) that divide the header and random barcode sequence\n",
    "\twhile True:\n",
    "\t\ttry: \n",
    "\t\t\tname_f = six.next(file_input).strip()\n",
    "\t\t\tseq_f  = six.next(file_input).strip()\n",
    "\t\t\tif name_f.startswith(\"@\"):\t# for fastq files\n",
    "\t\t\t\tsix.next(file_input)\t# To make sure that the next round it starts at a header again in the next round\n",
    "\t\t\t\tsix.next(file_input)\t# The program doesn't care about the quality values\t\t\t\t\n",
    "\t\t\t\tif re.search(splitter,name_f):\t# If there is a random barcode sequence in the header of the forward sequence...\n",
    "\t\t\t\t\trandbarcode = re.split(splitter, name_f)[-1].split()[0].split(\"@\")[0]  # split again with @ in case these are joined header from paired end fastq data\n",
    "\t\t\telif name_f.startswith(\">\"): # for fasta files\n",
    "\t\t\t\tif re.search(splitter,name_f):\t# If there is a random barcode sequence in the header of the forward sequence...\n",
    "\t\t\t\t\trandbarcode = re.split(splitter, name_f)[-1].split()[0].split(\">\")[0]  # split again with > in case these are joined header from paired end fasta data\n",
    "\t\t\t\n",
    "\t\t\ttracker[randbarcode,seq_f] += 1\n",
    "\t\t\t\t\t\n",
    "\t\texcept StopIteration:\n",
    "\t\t\tbreak\n",
    "\t\n",
    "\treturn tracker\n",
    "\n",
    "def trackPairedEndDuplicates(forward,reverse):\n",
    "\t\"\"\" Tracks duplicates in Paired-End data. Returns a dictionary \"\"\"\n",
    "\trandbarcode = None\n",
    "\theader\t\t= str()\n",
    "\ttracker\t\t= defaultdict(int)\n",
    "\tsplitter\t= \"_\"\t# the character(s) that divide the header and random barcode sequence\n",
    "\twhile True:\n",
    "\t\ttry:\n",
    "\t\t\tname_f, name_r\t= six.next(forward).strip(),six.next(reverse).strip()\n",
    "\t\t\tseq_f,\tseq_r\t= six.next(forward).strip(),six.next(reverse).strip()\n",
    "\t\t\tif name_f.startswith(\"@\"):\n",
    "\t\t\t\tsix.next(forward)\t\t# To make sure that the next round it starts at a header again in the next round\n",
    "\t\t\t\tsix.next(reverse)\t\t# The program doesn't care about the quality values\n",
    "\t\t\t\tsix.next(forward)\n",
    "\t\t\t\tsix.next(reverse)\n",
    "\t\t\tif re.search(splitter,name_f):\t# If there is a random barcode sequence in the header of the forward sequence...\n",
    "\t\t\t\trandbarcode = re.split(splitter, name_f)[-1].split()[0].split(\"@\")[0]\n",
    "\t\t\telse:\n",
    "\t\t\t\trandbarcode = \"##\"\n",
    "\t\t\ttracker[randbarcode,seq_f,seq_r] += 1\n",
    "\t\t\t\n",
    "\t\texcept StopIteration:\n",
    "\t\t\tbreak\n",
    "\t\n",
    "\treturn tracker\n",
    "\t\n",
    "def printFastaFile(tracker,output=None):\n",
    "\t\"\"\" prints the results in fasta format. Header includes random barcode sequence and number of duplicates \"\"\"\n",
    "\tif output:\n",
    "\t\tfile_output = open(output,\"w\")\n",
    "\telse:\n",
    "\t\tfile_output = sys.stdout\t\t# if it doesn't get\n",
    "\t\t\n",
    "\tfor rank,key in enumerate(tracker,start=1):\n",
    "\t\trandbarcode,sequence = key\n",
    "\t\tif randbarcode:\n",
    "\t\t\tfile_output.write(\">%s_%s_%s\\n%s\\n\" % (rank,randbarcode,tracker[randbarcode,sequence],sequence))\n",
    "\t\telse:\n",
    "\t\t\tfile_output.write(\">%s_%s\\n%s\\n\" % (rank,tracker[randbarcode,sequence],sequence))\n",
    "\n",
    "def printPairedEndFastaFile(tracker,output=None):\n",
    "\t\"\"\" prints the results in fasta format for Paired-End data. Header includes random barcode sequence and number of duplicates \"\"\"\n",
    "\tfile_name = str()\n",
    "\tif output:\n",
    "\t\tfile_name = getfilename(output)\n",
    "\telse:\n",
    "\t\tfile_name = \"collapsed\"\n",
    "\toutput_f = open(\"%s_1.fasta\" % (file_name), \"w\")\n",
    "\toutput_r = open(\"%s_2.fasta\" % (file_name), \"w\")\n",
    "\tfor rank,key in enumerate(tracker,start=1):\n",
    "\t\trandbarcode,sequence_f,sequence_r = key\n",
    "\t\tif randbarcode:\n",
    "\t\t\toutput_f.write(\">%s_%s_%s/1\\n%s\\n\" % (rank,randbarcode,tracker[randbarcode,sequence_f,sequence_r],sequence_f))\n",
    "\t\t\toutput_r.write(\">%s_%s_%s/2\\n%s\\n\" % (rank,randbarcode,tracker[randbarcode,sequence_f,sequence_r],sequence_r))\n",
    "\t\telse:\n",
    "\t\t\toutput_f.write(\">%s_%s/1\\n%s\\n\" % (rank,tracker[randbarcode,sequence_f,sequence_r],sequence_f))\n",
    "\t\t\toutput_r.write(\">%s_%s/2\\n%s\\n\" % (rank,tracker[randbarcode,sequence_f,sequence_r],sequence_r))\n",
    "\n",
    "def main():\n",
    "\tparser = OptionParser(usage=\"usage: %prog [options] -f forwardfile -r reversefile -o collapsed.fasta\", version=\"%s\" % __version__)\n",
    "\tfiles = OptionGroup(parser,\"File input options\")\n",
    "\tparser.add_option(\"-f\", \"--input_file\", dest=\"infile_forward\",help=\"name of the FASTQ or FASTA input file. Default is standard input\", metavar=\"FILE\",default=None)\n",
    "\tparser.add_option(\"-r\", \"--reverse_input_file\", dest=\"infile_reverse\",help=\"name of the paired (or reverse) FASTQ or FASTA input file\", metavar=\"FILE\",default=None)\n",
    "\tparser.add_option(\"-o\", \"--output_file\", dest=\"outfile\",help=\"Provide the path and name of the fasta output file. Default is standard output. For paired-end data just provide a file name without file extension (!)\",metavar=\"FILE\",default=None)\n",
    "\t(options, args) = parser.parse_args()\n",
    "\tinfile_f  = sys.stdin\n",
    "\toutfile\t  = None\n",
    "\tif options.outfile: \n",
    "\t\toutfile = options.outfile\n",
    "\tif options.infile_forward and not options.infile_reverse: \n",
    "\t\tinfile_f = open(options.infile_forward,\"r\")\n",
    "\t\tdatadict = trackDuplicates(infile_f)\n",
    "\t\tprintFastaFile(datadict,outfile)\n",
    "\telif options.infile_forward and options.infile_reverse: \n",
    "\t\tinfile_f = open(options.infile_forward,\"r\")\n",
    "\t\tinfile_r = open(options.infile_reverse,\"r\")\n",
    "\t\tdatadict = trackPairedEndDuplicates(infile_f,infile_r)\n",
    "\t\tprintPairedEndFastaFile(datadict,outfile)\n",
    "\telse:\n",
    "\t\tdatadict = trackDuplicates(infile_f)\n",
    "\t\tprintFastaFile(datadict,outfile)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "\tmain()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8ff3f1fb-be8a-4387-b357-4bcfaaf86c1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "tracker = trackPairedEndDuplicates(f, r)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "45a0ec57-46ef-4a32-82ec-277c8302d9a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "printPairedEndFastaFile(tracker, output = \"/Volumes/Extreme SSD/HP/read.dedup.fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b06519c-4c7a-428d-9073-b76cfc438e4f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
